src/config.rs

```diff
@@ -0,0 +1,376 @@
+// limitations under the License.
+use miette::{bail, ensure, miette, Context, IntoDiagnostic};
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+use crate::{
+    manifest::{Edition, CANARY_EDITION},
+    registry::{RegistryRef, RegistryUri},
+};
+
+/// Location of the configuration file
+const CONFIG_FILE: &str = ".buffrs/config.toml";
+
+/// Key for common default arguments
+const DEFAULT_ARGS_KEY: &str = "*";
+
+/// Representation of the configuration file
+///
+/// # Example
+///
+/// ```toml
+/// edition = "0.10.0"
+///
+/// [registries]
+/// some_org = "https://artifactory.example.com/artifactory/some-org"
+///
+/// [registry]
+/// default = "some_org"
+///
+/// [commands]
+/// default_args = ["--insecure"]
+///
+/// [commands.install]
+/// default_args = ["--generate-buf-yaml"]
+/// ```
+///
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Config {
+    /// Edition of this configuration file (in sync with the Proto.toml edition)
+    edition: Edition,
+
+    /// Path to the configuration file
+    config_path: Option<PathBuf>,
+
+    /// Default registry alias to use if none is specified
+    default_registry: Option<String>,
+
+    /// List of registries
+    registries: HashMap<String, RegistryUri>,
+
+    /// Default arguments for commands
+    command_defaults: HashMap<String, Vec<String>>,
+}
+
+impl Config {
+    /// Create a new configuration with default values
+    /// # Arguments
+    /// * `cwd` - Starting directory to search for the configuration file
+    ///
+    pub fn new(cwd: Option<&Path>) -> miette::Result<Self> {
+        match Self::locate_config(cwd) {
+            Some(config_path) => Self::new_from_config_file(&config_path),
+            None => Ok(Self {
+                edition: Edition::latest(),
+                config_path: None,
+                default_registry: None,
+                registries: HashMap::new(),
+                command_defaults: HashMap::new(),
+            }),
+        }
+    }
+
+    /// Parse a registry argument
+    ///
+    /// # Arguments
+    /// * `registry` - The registry argument to parse
+    ///
+    /// # Returns
+    /// URI with either alias scheme or actual URI:
+    /// - <alias> -> alias://<alias>
+    /// - <uri> -> <uri>
+    /// - None -> alias://<default>
+    pub fn parse_registry_arg(&self, registry: &Option<String>) -> miette::Result<RegistryRef> {
+        match registry {
+            Some(registry) => RegistryRef::from_str(registry),
+            None => match &self.default_registry {
+                Some(default_registry) => RegistryRef::from_str(default_registry),
+                None => bail!("no registry provided and no default registry found"),
+            },
+        }
+    }
+
+    /// Lookup a registry by name
+    ///
+    /// # Arguments
+    /// * `name` - Name of the registry to lookup
+    ///
+    /// # Returns
+    /// The registry URI
```

Comment:
Would you eventually remove the doc comments argument and return type bodies? They will get out of sync and rust doc does a great job at this

Reply:
If you wish. But rustdoc won't tell me that `name` is a registry name to look up. A newtype of course will.